### PR TITLE
Preserve torch._check messages in runtime asserts

### LINKED ISSUE #150063: Torch.check does not preserve original error message in the deferred runtime assert
```

@torch.compile(
    backend="inductor",
    fullgraph=True,
    dynamic=True,
)
def f(a, b):
    torch._check(
            a.size()[0] > 0,
            "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
            "because the operation does not have an identity",
        )
    # _check_vector_norm_args(a, -1, 1)

    return b*10

a = torch.ones(10, 10, device="cuda",dtype=torch.float64)
b = torch.torch.ones(10, 10, device="cuda",dtype=torch.float64) 
torch._dynamo.decorators.mark_unbacked(a, 0)
f(a, b)
```
output:
```
 /home/lsakka/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "Sym(u0)", arg1_1: "Sym(s0)", arg2_1: "f64[s0, s0][s0, 1]cuda:0"):
         # File: /home/lsakka/pytorch/example8.py:56 in f, code: a.size()[0] > 0,
        ge_1: "Sym(u0 >= 0)" = arg0_1 >= 0
        _assert_scalar = torch.ops.aten._assert_scalar.default(ge_1, "Runtime assertion failed for expression u0 >= 0 on node 'ge'");  ge_1 = _assert_scalar = None
        
        # No stacktrace found for following nodes
        gt: "Sym(u0 > 0)" = arg0_1 > 0;  arg0_1 = None
        _assert_scalar_1 = torch.ops.aten._assert_scalar.default(gt, "Runtime assertion failed for expression 0 < u0 on node 'gt_1'");  gt = _assert_scalar_1 = None
        
         # File: /home/lsakka/pytorch/example8.py:62 in f, code: return b*10
        mul: "f64[s0, s0][s0, 1]cuda:0" = torch.ops.aten.mul.Tensor(arg2_1, 10);  arg2_1 = None
        return (mul,)
        
```
instead of "Runtime assertion failed for expression 0 < u0 on node 'gt_1'" we should get 
 "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
            "because the operation does not have an identity",

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185734

torch._check already accepts lazy message callables, but the symbolic expect_true path discarded them before creating deferred runtime asserts. As a result, export and compile graphs emitted only the generated symbolic expression text for aten._assert_scalar, which hid the original framework error message requested by the check.

Thread the optional message through expect_true, SymNode, ShapeEnv RuntimeAssert metadata, and the SymBool/SymNode C++ bindings. Runtime assert insertion now uses the preserved message when available, including for derived single-symbol bound asserts that can be emitted before the original expression assert. The implementation keeps raw callables out of ShapeEnv cache/event keys, evaluates them only when a RuntimeAssert is actually recorded, and stores rendered strings for ShapeEnv event replay.

The alternative of only changing runtime_assert.py was not sufficient because the message had already been dropped before RuntimeAssert construction. Eagerly rendering the callable before the cached ShapeEnv boundary was also rejected because it regressed torch._check laziness and unhashable callable messages.

Fixes #150063

Generated by my agent

Test Plan:

python test/test_dynamic_shapes.py TestPySymInt.test_expect_true_message_replays TestPySymInt.test_expect_true_cpp_symnode_keeps_message_lazy TestPySymInt.test_runtime_assert_message_does_not_patch_parent_event

python test/export/test_export.py TestExport.test_torch_check_deferred_runtime_assert_preserves_message TestExport.test_python_asserts_with_sym_int TestExport.test_dim_dynamic TestExport.test_export_mod_constraints

lintrunner -a

python - <<'PY' ... torch.compile(backend=inspection_backend, fullgraph=True, dynamic=True) ... mark_unbacked(a, 0) ... PY

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR after the CI failures. The test failures were expectations still looking for the old generic runtime assert text; they now expect the preserved original reshape/vector_norm messages, and the docs coverage failure is addressed by documenting ShapeEnv.guard_or_defer_runtime_assert.

**Comment by jansel** [comment]
My agent says I updated the PR for the remaining Dr. CI failures. The failures were stale expectations in unbacked expand/reshape tests after this PR started preserving the original operator error messages; the tests now expect the preserved non-broadcasting and invalid-shape messages. Targeted affected tests and lintrunner -a pass locally.

**Comment by jansel** [comment]
My agent says I updated the PR for the latest Dr. CI failure in TestExportCustomClass.test_unbacked_contiguous. The test was still expecting the old generic runtime assert strings for reshape checks; it now expects the preserved reshape error messages produced by this PR. The failing test and related export/dynamic-shapes targets pass locally.

**Comment by jansel** [comment]
My agent says I updated the PR for the retraceability CI failures. Retracing rebuilt deferred runtime asserts from existing aten._assert_scalar nodes but lost their original string messages, so the failing unbacked expand retrace test saw generic Eq text again. Runtime assert insertion now recovers existing assert messages by expression, including for normalized bound asserts, and the failing retraceability targets pass locally.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the runtime-assert retrace regression behind the latest Dr. CI failures. The change now avoids reusing stale generated assert-message text for derived bound asserts, while still preserving real custom/runtime assert messages. Targeted export, retrace, serdes, dynamic-shapes, passes, and control-flow tests pass locally, and lintrunner -a is clean.

**Comment by jansel** [comment]
My agent says I updated the PR for the latest Dr. CI failures where derived runtime bound asserts rendered fallback messages as expression True. Runtime assert message rendering now ignores folded BooleanAtom node metadata and falls back to the original symbolic RuntimeAssert or bound expression, so the assertions keep useful predicates like u0 >= 6. Targeted dynamic-shapes validation and lintrunner -a pass locally; further export test runs were blocked by local stuck torch import processes after the initial reproduction.

**Comment by jansel** [comment]
My agent says I updated the PR for the latest Dr. CI failures in test_no_check_is_size_error. The failing tests were only matching one textual ordering of an equivalent symbolic bound; the assertion now accepts both u >= 6 and 6 <= u renderings while still checking for the runtime assert message. The two failing export/retrace targets and lintrunner -a pass locally.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
